<h2>Why is this an issue?</h2>
<p>Use of the <code>exec</code> statement could be dangerous, and should be avoided. Moreover, the <code>exec</code> statement was removed in Python
3.0. Instead, the built-in <code>exec()</code> function can be used.</p>
<h3>Noncompliant code example</h3>
<pre>
exec 'print 1' # Noncompliant
</pre>
<h3>Compliant solution</h3>
<pre>
exec('print 1')
</pre>

