<h2>Why is this an issue?</h2>
<p>The forms <code>&lt;&gt;</code> and <code>!=</code> are equivalent. But in Python 2.7.3 the <code>&lt;&gt;</code> form is considered obsolete.</p>
<h3>Noncompliant code example</h3>
<pre>
return a &lt;&gt; b # Noncompliant
</pre>
<h3>Compliant solution</h3>
<pre>
return a != b
</pre>

