<h2>Why is this an issue?</h2>
<p>The long suffix should always be written in uppercase, i.e. 'L', as the lowercase 'l' can easily be confused with the digit one '1'.</p>
<h3>Noncompliant code example</h3>
<pre>
return 10l  // Noncompliant; easily confused with one zero one
</pre>
<h3>Compliant solution</h3>
<pre>
return 10L
</pre>

