<h2>Why is this an issue?</h2>
<p>Any issue to quality rule can be deactivated with the <code>NOSONAR</code> marker. This marker is pretty useful to exclude false-positive results
but it can also be used abusively to hide real quality flaws.</p>
<p>This rule raises an issue when <code>NOSONAR</code> is used.</p>
<h3>Noncompliant code example</h3>
<pre>
for d in lib_dirs:
    # NOSONAR: lib_dirs is undefined
    pass
</pre>

