<p>This rule raises an issue when a pre/post increment or decrement operator is used.</p>
<h2>Why is this an issue?</h2>
<p>Python has no pre/post increment/decrement operator. For instance, <code>x++</code> and <code>x--</code> will fail to parse. More importantly,
<code>++x</code> and <code>--x</code> will do nothing. To increment a number, simply write <code>x += 1</code>.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
++x # Noncompliant: pre and post increment operators do not exist in Python.

x-- # Noncompliant: pre and post decrement operators do not exist in Python.
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
x += 1

x -= 1
</pre>

