<h2>Why is this an issue?</h2>
<p>The <code>print</code> statement was removed in Python 3.0. The built-in function should be used instead.</p>
<h3>Noncompliant code example</h3>
<pre>
print '1'  # Noncompliant
</pre>
<h3>Compliant solution</h3>
<pre>
print('1')
</pre>

