<p>This rule raises an issue when an <code>except</code> statement has had all its exceptions caught by a previous <code>except</code> clause.</p>
<h2>Why is this an issue?</h2>
<p>Exceptions handlers (<code>except</code>) are evaluated in the order they are written. Once a match is found, the evaluation stops.</p>
<p>In some contexts, an except block is dead code as it will never catch any exception:</p>
<ul>
  <li> If there is a handler for a base class followed by a handler for class derived from that base class, the second handler will never trigger: The
  handler for the base class will match the derived class, and will be the only executed handler. </li>
  <li> When multiple <code>except</code> statements try to catch the same exception class, only the first one will be executed. </li>
  <li> In Python 3, <code>BaseException</code> is the parent of every exception class. When a <code>BaseException</code> is caught by an
  <code>except</code> clause, none of the subsequent <code>except</code> statement will catch anything. This is true as well for the bare except
  statement (<code>except:</code>). </li>
</ul>
<h2>How to fix it</h2>
<p>When using multiple <code>except</code> statements, make sure to:</p>
<ul>
  <li> Order the <code>except</code> blocks from the most specialzed exception to the most generic, i.e when wanting to catch a
  <code>FloatingPointError</code> and an <code>ArithemticError</code>, as <code>FloatingPointError</code> is a subclass of
  <code>ArithmeticError</code>, the first <code>except</code> statement should be <code>FloatingPointError</code>. </li>
  <li> Catch the same exception only once. </li>
  <li> Catch a <code>BaseException</code> only once with either an <code>except BaseException:</code> statement or a bare <code>except:</code>
  statement, as the two statements are equivalent. </li>
</ul>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
def foo():
    try:
        raise FloatingPointError()
    except (ArithmeticError, RuntimeError) as e:
        print(e)
    except FloatingPointError as e: # Noncompliant: FloatingPointError is a subclass of ArithmeticError.
        print("Never executed")
    except OverflowError as e: # Noncompliant: OverflowError is a subclass of ArithmeticError.
        print("Never executed")

    try:
        raise TypeError()
    except TypeError as e:
        print(e)
    except TypeError as e: # Noncompliant: duplicate except.
        print("Never executed")

    try:
        raise ValueError()
    except BaseException as e:
        print(e)
    except: # Noncompliant: this is equivalent to "except BaseException" block.
        print("Never executed")
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
def foo():
    try:
        raise FloatingPointError()
    except FloatingPointError as e:
        print("Executed")
    except OverflowError as e:
        print("Executed")
    except (ArithmeticError, RuntimeError) as e:
        print(e)

    try:
        raise TypeError()
    except TypeError as e:
        print(e)

    try:
        raise ValueError()
    except BaseException as e:
        print(e)
</pre>
<p><strong>Note</strong>: <em>It is generally not recommended to try catching <code>BaseException</code>, as it is the base class for all built-in
exceptions in Python, including system-exiting exceptions like <code>SystemExit</code> or <code>KeyboardInterrupt</code>, which are typically not
meant to be caught. See <a href="https://www.python.org/dev/peps/pep-0352/#exception-hierarchy-changes">PEP 352</a> for more information.</em></p>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> <a href="https://docs.python.org/3/reference/compound_stmts.html#the-try-statement">The <code>try</code> statement</a> </li>
  <li> <a href="https://docs.python.org/3/library/exceptions.html#exception-hierarchy">Exception hierarchy</a> </li>
</ul>

