<h2>Why is this an issue?</h2>
<p>Most of the time a block of code is empty when a piece of code is really missing. So such empty block must be either filled or removed.</p>
<h3>Noncompliant code example</h3>
<pre>
for i in range(3):
    pass
</pre>
<h3>Exceptions</h3>
<p>When a block contains a comment, this block is not considered to be empty.</p>

