<p>Unused parameters are misleading. Whatever the value passed to such parameters is, the behavior will be the same.</p>
<h2>Noncompliant Code Example</h2>
<pre>
def do_something(a, b): # "b" is unused
  return compute(a)
</pre>
<h2>Compliant Solution</h2>
<pre>
def do_something(a):
  return compute(a)
</pre>
<h2>Exceptions</h2>
<p>Overriding methods are ignored.</p>
<pre>
class C(B):
  def do_something(self, a, b): # no issue reported on b
    return self.compute(a)
}
</pre>
<p>Throwaway variables <code>_</code>.</p>
<pre>
def do_something(a, _): # no issue reported on _
  return compute(a)
</pre>

