<p>There are several reasons for a function or a method not to have a body:</p>
<ul>
  <li> It is an unintentional omission, and should be fixed to prevent an unexpected behavior in production. </li>
  <li> It is not yet, or never will be, supported. In this case an exception should be thrown. </li>
  <li> The method is an intentionally-blank override. In this case a nested comment should explain the reason for the blank override. </li>
</ul>
<h2>Noncompliant Code Example</h2>
<pre>
def myfunc1():  # Noncompliant
    pass

def myfunc2():  # Noncompliant
    """
    this is a docstring
    """

def myfunc3():
    pass  # comment explaining why this function is empty


class MyClass:
    def mymethod1(self):  # Noncompliant
        pass

    def mymethod3(self):  # Noncompliant
        """
        this is a docstring
        """
</pre>
<h2>Compliant Solution</h2>
<pre>
def myfunc1():
    """
    this is a docstring
    """
    pass  # comment explaining why this function is empty

def myfunc2():
    """
    this is a docstring
    """
    raise NotImplementedError()

class MyClass:
    def mymethod1(self):
        """
        this is a docstring
        """
        pass  # comment explaining why this function is empty

    def mymethod2(self):
        """
        this is a docstring
        """
        raise NotImplementedError()
</pre>
<h2>Exceptions</h2>
<p>No issue will be raised when the empty method is abstract and meant to be overriden in a subclass, i.e. it is annotated with
<code>abc.abstractmethod</code>, <code>abc.abstractstaticmethod</code>, <code>abc.abstractclassmethod</code> or <code>abc.abstractproperty</code></p>
<pre>
import abc

class MyAbstractClass(abc.ABC):
    @abc.abstractproperty
    def myproperty(self):
        pass

    @abc.abstractclassmethod
    def myclassmethod(cls):
        pass

    @abc.abstractmethod
    def mymethod(self):
        pass

    @abc.abstractstaticmethod
    def mystaticmethod():
        pass
</pre>

