<h2>Why is this an issue?</h2>
<p>If a local variable is declared but not used, it is dead code and should be removed. Doing so will improve maintainability because developers will
not wonder what the variable is used for.</p>
<h3>Noncompliant code example</h3>
<pre>
def hello(name):
    message = "Hello " + name # Noncompliant
    print(name)
for i in range(10):
    foo()
</pre>
<h3>Compliant solution</h3>
<pre>
def hello(name):
    message = "Hello " + name
    print(message)
for _ in range(10):
    foo()
</pre>
<h3>Exceptions</h3>
<p><code>_</code> will not raise an issue for this rule. The following examples are compliant:</p>
<pre>
for _ in range(10):
    do_something()
username, login, _ = auth
do_something_else(username, login)
</pre>

