<p>Function arguments should all have different names to prevent any ambiguity. Indeed, if arguments have the same name, the last duplicated argument
hides all the previous arguments with the same name. This hiding makes no sense, reduces understandability and maintainability, and obviously can be
error prone. </p>
<h2>Noncompliant Code Example</h2>
<pre>
def other(a, a, c): # Noncompliant
    [...]
    return a * c
</pre>
<h2>Compliant Solution</h2>
<pre>
def other(a, c): # Compliant
    [...]
    return a * c
</pre>

