<h2>Why is this an issue?</h2>
<p>Shared coding conventions allow teams to collaborate efficiently. This rule checks that all function names match a provided regular expression.</p>
<h3>Noncompliant code example</h3>
<p>With the default provided regular expression: <code>^[a-z_][a-z0-9_]*$</code></p>
<pre>
def MyFunction(a,b):
    ...
</pre>
<h3>Compliant solution</h3>
<pre>
def my_function(a,b):
    ...
</pre>

