<p>Shared coding conventions allow teams to collaborate efficiently. This rule checks that all function names match a provided regular expression.</p>
<h2>Noncompliant Code Example</h2>
<p>With the default provided regular expression: <code>^[a-z_][a-z0-9_]*$</code></p>
<pre>
def MyFunction(a,b):
    ...
</pre>
<h2>Compliant Solution</h2>
<pre>
def my_function(a,b):
    ...
</pre>

