<p>This rule is deprecated, and will eventually be removed.</p>
<h2>Why is this an issue?</h2>
<p>Typically, backslashes are seen only as part of escape sequences. Therefore, the use of a backslash outside of a raw string or escape sequence
looks suspiciously like a broken escape sequence.</p>
<p>Characters recognized as escape-able are: <code>abfnrtvox\'"</code></p>
<h3>Noncompliant code example</h3>
<pre>
s = "Hello \world."
t = "Nice to \ meet you"
u = "Let's have \ lunch"
</pre>
<h3>Compliant solution</h3>
<pre>
s = "Hello world."
t = "Nice to \\ meet you"
u = r"Let's have \ lunch"  // raw string
</pre>

