<p>This rule is deprecated; use {rule:python:S5905} instead.</p>
<h2>Why is this an issue?</h2>
<p>Parentheses are not required after the <code>assert</code>, <code>del</code>, <code>elif</code>, <code>except</code>, <code>for</code>,
<code>if</code>, <code>in</code>, <code>not</code>, <code>raise</code>, <code>return</code>, <code>while</code>, and <code>yield</code> keywords, and
using them unnecessarily impairs readability. They should therefore be omitted.</p>
<h3>Noncompliant code example</h3>
<pre>
x = 1
while (x &lt; 10):
    print "x is now %d" % (x)
    x += 1
</pre>
<h3>Compliant solution</h3>
<pre>
x = 1
while x &lt; 10:
    print "x is now %d" % (x)
    x += 1
</pre>

