<p>A loop with at most one iteration is equivalent to the use of an <code>if</code> statement to conditionally execute one piece of code. No developer
expects to find such a use of a loop statement. If the initial intention of the author was really to conditionally execute one piece of code, an
<code>if</code> statement should be used instead.</p>
<p>At worst that was not the initial intention of the author and so the body of the loop should be fixed to use the nested <code>return</code>,
<code>break</code> or <code>raise</code> statements in a more appropriate way.</p>
<h2>Noncompliant Code Example</h2>
<pre>
while node is not None:
  node = node.parent()
  print(node)
  break
</pre>
<h2>Compliant Solution</h2>
<pre>
while node is not None:
  node = node.parent()
  print(node)
</pre>

