<h2>Why is this an issue?</h2>
<p>Looking at the set of methods and fields in a <code>class</code> and finding two that differ only by capitalization is confusing to users of the
class.</p>
<p>This situation may simply indicate poor naming. Method names should be action-oriented, and thus contain a verb, which is unlikely in the case
where both a method and a field have the same name (with or without capitalization differences). However, renaming a public method could be disruptive
to callers. Therefore renaming the member is the recommended action.</p>
<h3>Noncompliant code example</h3>
<pre>
class SomeClass:
    lookUp = false
    def lookup():       # Non-compliant; method name differs from field name only by capitalization
        pass
</pre>
<h3>Compliant solution</h3>
<pre>
class SomeClass:
    lookUp = false
    def getLookUp():
        pass
</pre>

