<p>This rule raises an issue when there is a method and a field in a class with names that differ only by capitalization.</p>
<h2>Why is this an issue?</h2>
<p>Looking at the set of methods and fields in a <code>class</code> and finding two that differ only by capitalization is confusing to users of the
class.</p>
<p>This situation may simply indicate poor naming. Method names should be action-oriented, and thus contain a verb, which is unlikely in the case
where both a method and a field have the same name (with or without capitalization differences). However, renaming a public method could be disruptive
to callers. Therefore renaming the member is the recommended action.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
class SomeClass:
    lookUp = false
    def lookup():       # Non-compliant; method name differs from field name only by capitalization
        pass
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
class SomeClass:
    lookUp = false
    def getLookUp():
        pass
</pre>

