<p>Having two branches in the same <code>if</code> structure with the same implementation is at best duplicate code, and at worst a coding error. If
the same logic is truly needed for both instances, then they should be combined.</p>
<h2>Noncompliant Code Example</h2>
<pre>
if 0 &lt;= a &lt; 10:
    do_first()
    do_second()
elif 10 &lt;= a &lt; 20:
    do_the_other_thing()
elif 20 &lt;= a &lt; 50:
    do_first()         # Noncompliant; duplicates first condition
    do_second()
</pre>
<h2>Exceptions</h2>
<p>Blocks in an <code>if</code> chain that contain a single line of code are ignored.</p>
<pre>
if 0 &lt;= a &lt; 10:
    do_first()
elif 10 &lt;= a &lt; 20:
    do_the_other_thing()
elif 20 &lt;= a &lt; 50:
    do_first()         # no issue, usually this is done on purpose to increase the readability
</pre>
<p>But this exception does not apply to <code>if</code> chains without <code>else</code>-s when all branches have the same single line of code. In
case of <code>if</code> chains with <code>else</code>-s rule {rule:python:S3923} raises a bug. </p>
<pre>
if 0 &lt;= a &lt; 10:
    do_first()
elif 20 &lt;= a &lt; 50:
    do_first()         # Noncompliant, this might have been done on purpose but probably not
</pre>

