<h2>Why is this an issue?</h2>
<p><code>yield</code> and <code>return</code> only make sense in the context of functions. Using them outside a function raises a
<code>SyntaxError</code>. To break out of a loop, use <code>break</code> instead.</p>
<h3>Noncompliant code example</h3>
<pre>
class MyClass:
    while True:
        return False #Noncompliant
</pre>

