<p>This rule is deprecated; use {rule:python:S5722} instead.</p>
<h2>Why is this an issue?</h2>
<p>The <code>__exit__</code> method is invoked with four arguments: self, type, value and traceback. Leave one of these out of the method declaration
and the result will be a <code>TypeError</code> at runtime.</p>
<h3>Noncompliant code example</h3>
<pre>
class MyClass:
   def __enter__(self):
       pass
   def __exit__(self, exc_type, exc_val):  # Noncompliant
       pass
</pre>

