<p>This rule raises an issue when the <code>__init__</code> method of a class contains a <code>return</code> or a <code>yield</code> statement.</p>
<h2>Why is this an issue?</h2>
<p>By contract, every Python function returns something, even if it is the <code>None</code> value, which can be returned implicitly by omitting the
<code>return</code> statement, or explicitly.</p>
<p>The <code>__init__</code> method is required to return <code>None</code>. A <code>TypeError</code> will be raised if the <code>__init__</code>
method either yields or returns any expression other than <code>None</code>. While explicitly returning an expression that evaluates to
<code>None</code> will not raise an error, it is considered bad practice.</p>
<p>To fix this issue, make sure that the <code>__init__</code> method does not contain any return statement.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
class MyClass(object):
    def __init__(self):
        self.message = 'Hello'
        return self  # Noncompliant: a TypeError will be raised
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
class MyClass(object):
    def __init__(self):
        self.message = 'Hello'
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> <a href="https://docs.python.org/3/reference/datamodel.html#object.__init__">The <code>__init__</code> method</a> </li>
</ul>

