<h2>Why is this an issue?</h2>
<p>An <code>except</code> clause that only rethrows the caught exception has the same effect as omitting the <code>except</code> altogether and
letting it bubble up automatically, but with more code and the additional detriment of leaving maintainers scratching their heads.</p>
<p>Such clauses should either be eliminated or populated with the appropriate logic.</p>
<h3>Noncompliant code example</h3>
<pre>
a = {}
try:
    a[5]
except KeyError:
    raise  # Noncompliant
</pre>
<h3>Compliant solution</h3>
<pre>
a = {}
try:
    a[5]
except KeyError as e:
    logging.exception('error while accessing the dict')
    raise e
</pre>

