<p>Calling the <code>not</code> or <code>~</code> prefix operator twice might be redundant: the second invocation undoes the first. Such mistakes are
typically caused by accidentally double-tapping the key in question without noticing. Either this is a bug, if the operator was actually meant to be
called once, or misleading if done on purpose.</p>
<h2>Noncompliant Code Example</h2>
<pre>
a = 0
b = False

c = not not a # Noncompliant
d = ~~b # Noncompliant
</pre>
<h2>Compliant Solution</h2>
<pre>
a = 0
b = False

c = not a # Compliant
d = ~b # Compliant
</pre>

