<h2>Why is this an issue?</h2>
<p>The use of a <code>pass</code> statement where itâ€™s not required by the syntax is pure cruft and should be removed.</p>
<h3>Noncompliant code example</h3>
<pre>
    def __init__(self, log="", who="", date=0, files=[]):
        self.log = log
        self.files = files
        self.who = who
        self.date = date
        pass    # Noncompliant

    def lookup():
        pass    # Compliant; method can't be empty
</pre>
<h3>Compliant solution</h3>
<pre>
    def __init__(self, log="", who="", date=0, files=[]):
        self.log = log
        self.files = files
        self.who = who
        self.date = date

    def lookup():
        pass
</pre>

