<h2>Why is this an issue?</h2>
<p><code>__all__</code> is used to define the list of moduleâ€™s names that should be imported when <code>from package import *</code> is used. For that
reason, it may only contain strings.</p>
<h3>Noncompliant code example</h3>
<pre>
class MyClass:
    pass

__all__ = [
    MyClass  # Noncompliant
]
</pre>
<h3>Compliant solution</h3>
<pre>
class MyClass:
    pass

__all__ = [
    "MyClass"
]
</pre>
<h2>Resources</h2>
<ul>
  <li> <a href="https://docs.python.org/3/tutorial/modules.html#importing-from-a-package">Python documentation - Importing * From a Package</a> </li>
</ul>

