<p>Just because you <em>can</em> do something, doesn't mean you should, and that's the case with nested conditional expressions. Nesting conditional
expressions results in the kind of code that may seem clear as day when you write it, but six months later will leave maintainers (or worse - future
you) scratching their heads and cursing.</p>
<p>Instead, err on the side of clarity, and use another line to express the nested operation as a separate statement.</p>
<h2>Noncompliant Code Example</h2>
<pre>
class Job:
    @property
    def readable_status(self):
        return "Running" if job.is_running else "Failed" if job.errors else "Succeeded"  # Noncompliant
</pre>
<h2>Compliant Solution</h2>
<pre>
class Job:
    @property
    def readable_status(self):
        if job.is_running:
            return "Running"
        return "Failed" if job.errors else "Succeeded"
</pre>
<h2>Exceptions</h2>
<p>No issue is raised on conditional expressions in comprehensions.</p>

