<p>This rule raises an issue when a function call result is used, even though the function does not return anything.</p>
<h2>Why is this an issue?</h2>
<p>If a function does not return anything, it makes no sense to use its output. Specifically, passing it to another function, or assigning its
"result" to a variable is probably a bug because such functions return nothing, which is probably not what was intended.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
def foo():
  l = [1, 2, 3]
  result = l.append(42) # Noncompliant, `append` mutates list l
  return result
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
def foo():
  l = [1, 2, 3]
  l.append(42)
  return l
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Python documentation - <a href="https://docs.python.org/3/tutorial/controlflow.html#defining-functions">Defining Functions</a> </li>
</ul>

