<h2>Why is this an issue?</h2>
<p>Variables, Classes and functions should be defined before they are used, otherwise the code will fail.</p>
<h3>Noncompliant code example</h3>
<pre>
def noncompliant():
    foo()  # Noncompliant
    foo = sum

    func()  # Noncompliant
    def func():
        pass

    MyClass()  # Noncompliant
    class MyClass:
        pass
</pre>
<h3>Compliant solution</h3>
<pre>
def compliant():
    foo = sum
    foo()

    def func():
        pass
    func()

    class MyClass:
        pass
    MyClass()
</pre>

