<p>Variables, Classes and functions should be defined before they are used, otherwise the code will fail.</p>
<h2>Noncompliant Code Example</h2>
<pre>
def noncompliant():
    foo()  # Noncompliant
    foo = sum

    func()  # Noncompliant
    def func():
        pass

    MyClass()  # Noncompliant
    class MyClass:
        pass
</pre>
<h2>Compliant Solution</h2>
<pre>
def compliant():
    foo = sum
    foo()

    def func():
        pass
    func()

    class MyClass:
        pass
    MyClass()
</pre>

