<p>An <code>re.sub</code> call always performs an evaluation of the first argument as a regular expression, even if no regular expression features
were used. This has a significant performance cost and therefore should be used with care.</p>
<p>When <code>re.sub</code> is used, the first argument should be a real regular expression. If it’s not the case, <code>str.replace</code> does
exactly the same thing as <code>re.sub</code> without the performance drawback of the regex.</p>
<p>This rule raises an issue for each <code>re.sub</code> used with a simple string as first argument which doesn’t contains special regex character
or pattern.</p>
<h2>Noncompliant Code Example</h2>
<pre>
input = "Bob is a Bird... Bob is a Plane... Bob is Superman!"
changed = re.sub("Bob is", "It's", input) # Noncompliant
changed = re.sub("\.\.\.", ";", changed) # Noncompliant
</pre>
<h2>Compliant Solution</h2>
<pre>
input = "Bob is a Bird... Bob is a Plane... Bob is Superman!"
changed = str.replace("Bob is", "It's", input)
changed = str.replace("...", ";", changed)
</pre>
<p>Or, with a regex:</p>
<pre>
input = "Bob is a Bird... Bob is a Plane... Bob is Superman!"
changed = re.sub(r"\w*\sis", "It's", input)
changed = re.sub(r"\.{3}", ";", changed)
</pre>

