<p>This rule raises an issue when an operator is used on incompatible types.</p>
<h2>Why is this an issue?</h2>
<p>For a specific operator, two types are considered incompatible if no built-in operations between those types exist and none of the operands has
implemented the operator’s corresponding special methods. Performing such an operation on incompatible types will raise a <code>TypeError</code>.</p>
<p>Calling an operator in Python is equivalent to calling a special method (except for the identity operator <code>is</code>). Python provides a set
of built-in operations. For example, to add two integers: <code>1 + 2</code>, calling the built-in operator <code>+</code> is equivalent to calling
the special method <code>__add__</code> on the type <code>int</code>.</p>
<p>Python allows developers to define how an operator will behave with a custom class by implementing the corresponding special method. When defining
such methods for symmetrical binary operators, developers need to define two methods so that the order of operands doesn’t matter, ex:
<code>__add__</code> and <code>__radd__</code>.</p>
<p>For a complete list of operators and their methods see the Python documentation: <a
href="https://docs.python.org/3/reference/datamodel.html?#emulating-numeric-types">arithmetic and bitwise operators</a>, <a
href="https://docs.python.org/3/reference/datamodel.html?#object.__lt__">comparison operators</a>.</p>
<h2>How to fix it</h2>
<p>Implementing the special methods for a specific operator will fix the issue.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre>
class Empty:
    pass

class Add:
    def __add__(self, other):
        return 42

Empty() + 1  # Noncompliant: no __add__ method is defined on the Empty class
Add() + 1
1 + Add()  # Noncompliant: no __radd__ method is defined on the Add class
Add() + Empty()
Empty() + Add()  # Noncompliant: no __radd__ method is defined on the Add class
</pre>
<h4>Compliant solution</h4>
<pre>
class Empty:
    pass

class Add:
    def __add__(self, other):
        return 42

    def __radd__(self, other):
        return 42

Add() + 1
1 + Add()
Add() + Empty()
Empty() + Add()
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> <a href="https://docs.python.org/3/reference/datamodel.html?#object.__lt__">Rich comparison methods</a> </li>
  <li> <a href="https://docs.python.org/3/reference/datamodel.html?#emulating-numeric-types">Emulating numeric types</a> </li>
</ul>

