<p>The CPython interpreter does not check arguments type when functions are called. However a function can express the type it expects for each
argument in its documentation or by using <a href="https://www.python.org/dev/peps/pep-0484/">Type Hints</a>. Calling such a function with an argument
of a different type can easily create a bug. Even if it works right now it can fail later when APIs evolve or when type checks are added (ex: with
<code>isinstance</code>).</p>
<p>This rule raises an issue when a function or method is called with an argument of a different type than the one described in its type annotations.
It also checks argument types for builtin functions.</p>
<h2>Noncompliant Code Example</h2>
<pre>
def func(var: str):
    pass

func(42)  # Noncompliant

len(1)  # Noncompliant
</pre>
<h2>Compliant Solution</h2>
<pre>
def func(var: str):
    pass

func("42")

len("1")
</pre>
<h2>See</h2>
<ul>
  <li> <a href="https://docs.python.org/3/library/functions.html#built-in-funcs">Python documentation - builtins</a> </li>
  <li> <a href="https://www.python.org/dev/peps/pep-0484/">PEP 484 — Type Hints</a> </li>
  <li> <a href="https://docs.python.org/3/library/typing.html">Python documentation - typing — Support for type hints</a> </li>
</ul>

