<p>Exception chaining enables users to see if an exception was triggered by another exception (see <a
href="https://www.python.org/dev/peps/pep-3134/">PEP-3134</a>). Exceptions are chained using either of the following syntax:</p>
<ul>
  <li> <code>raise NewException() from chained_exception</code> </li>
  <li> <code>new_exception.__cause__ = chained_exception</code> </li>
</ul>
<p>It is also possible to erase a chaining by setting <code>new_exception.__cause__ = None</code> or using <code>except ... from None</code> (see <a
href="https://www.python.org/dev/peps/pep-0409/">PEP-409</a>).</p>
<p>Chaining will fail and raise a <code>TypeError</code> if something else than <code>None</code> or a valid exception, i.e. an instance of
<code>BaseException</code> or of a subclass, is provided.</p>
<h2>Noncompliant Code Example</h2>
<pre>
class A:
    pass

try:
    raise ValueError("orig")
except ValueError as e:
    new_exc = TypeError("new")
    new_exc.__cause__ = A()  # Noncompliant
    raise new_exc

try:
    raise ValueError("orig")
except ValueError as e:
    raise TypeError("new") from "test"  # Noncompliant
</pre>
<h2>Compliant Solution</h2>
<pre>
try:
    raise ValueError("orig")
except ValueError as e:
    new_exc = TypeError("new")
    new_exc.__cause__ = None  # Ok
    raise new_exc

try:
    raise ValueError("orig")
except ValueError as e:
    new_exc = TypeError("new")
    new_exc.__cause__ = e  # Ok
    raise new_exc

try:
    raise ValueError("orig")
except ValueError as e:
    raise TypeError("new") from None  # Ok

try:
    raise ValueError("orig")
except ValueError as e:
    raise TypeError("new") from e  # Ok
</pre>
<h2>See</h2>
<ul>
  <li> PEP 3134 – <a href="https://www.python.org/dev/peps/pep-3134/">Exception Chaining and Embedded Tracebacks</a> </li>
  <li> PEP 409 – <a href="https://www.python.org/dev/peps/pep-0409/">Suppressing exception context</a> </li>
  <li> PEP 352 - <a href="https://www.python.org/dev/peps/pep-0352/#exception-hierarchy-changes">Required Superclass for Exceptions</a> </li>
  <li> Python documentation - <a href="https://docs.python.org/3/library/exceptions.html">Built-in Exceptions</a> </li>
</ul>

