<p>This rule raises an issue when the expression used in an <code>except</code> statement is neither a class deriving from <code>BaseException</code>
nor a tuple of such classes.</p>
<h2>Why is this an issue?</h2>
<p>In Python 3â€™s <code>except</code> statement, attempting to catch an object that does not derive from <code>BaseException</code> will raise a
<code>TypeError</code>.</p>
<p>In order to catch multiple exceptions in an <code>except</code> statement, a <code>tuple</code> of exception classes should be provided. Trying to
catch multiple exceptions with a <code>list</code> or a <code>set</code> will raise a <code>TypeError</code>.</p>
<p>If you are about to create a custom exception class, note that custom exceptions should inherit from <code>Exception</code>, rather than
<code>BaseException</code>.</p>
<p><code>BaseException</code> is the base class for all built-in exceptions in Python, including system-exiting exceptions like
<code>SystemExit</code> or <code>KeyboardInterrupt</code>, which are typically not meant to be caught. On the other hand, <code>Exception</code> is
intended for exceptions that are expected to be caught, which is generally the case for user-defined exceptions. See <a
href="https://www.python.org/dev/peps/pep-0352/#exception-hierarchy-changes">PEP 352</a> for more information.</p>
<p>To fix this issue, make sure the expression used in an <code>except</code> statement is an exception which derives from
<code>BaseException</code>/<code>Exception</code> or a tuple of such exceptions.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
class CustomException(object):
    """An Invalid exception class."""
    pass

try:
    ...
except CustomException:  # Noncompliant: this custom exception does not derive from BaseException or Exception.
    print("exception")

try:
    ...
except [TypeError, ValueError]:  # Noncompliant: list of exceptions, only tuples are valid.
    print("exception")
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
class CustomException(Exception):
    pass

try:
    ...
except CustomException:
    print("exception")

try:
    ...
except (TypeError, ValueError):
    print("exception")
</pre>
<p><strong>Note</strong>: <em>In Python 2 it is possible to raise and catch old-style classes, which do not inherit from
<code>BaseException</code>.</em></p>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Python Documentation - <a href="https://docs.python.org/3/tutorial/errors.html">Errors and Exceptions</a> </li>
  <li> Python Documentation - <a href="https://docs.python.org/3/reference/compound_stmts.html#except">The <code>try</code> statement</a> </li>
  <li> Python Documentation - <a href="https://docs.python.org/3/library/exceptions.html#exception-hierarchy">Exception hierarchy</a> </li>
</ul>
<h3>Standards</h3>
<ul>
  <li> PEP 352 - <a href="https://www.python.org/dev/peps/pep-0352/#exception-hierarchy-changes">Required Superclass for Exceptions</a> </li>
</ul>

