<p>Instance methods, i.e. methods not annotated with <code>@cassmethod</code> or <code>@staticmethod</code>, are expected to have at least one
parameter. This parameter will reference the object instance on which the method is called. By convention, this first parameter is named "self".</p>
<p>Naming the "self" parameter differently is confusing. It might also indicate that the "self" parameter was forgotten, in which case calling the
method will most probably fail.</p>
<p>This rule raises an issue when the first parameter of an instance method is not called "self".</p>
<h2>Noncompliant Code Example</h2>
<pre>
class MyClass:
    def send_request(request):  # Noncompliant. "self" was probably forgotten
        print("send_request")
</pre>
<h2>Compliant Solution</h2>
<pre>
class MyClass:
    def send_request(self, request):
        print("send_request")
</pre>
<h2>Exceptions</h2>
<p>No issue will be raised for methods called <code>__init_subclass__</code>, <code>__class_getitem__</code> or <code>__new__</code> as these methods'
first parameter is a class.</p>
<h2>See</h2>
<ul>
  <li> Python documentation - <a href="https://docs.python.org/3.8/tutorial/classes.html#method-objects">Method Objects</a> </li>
  <li> PEP8 - <a href="https://www.python.org/dev/peps/pep-0008/#function-and-method-arguments">Function and Method Arguments</a> </li>
</ul>

