<p>This rule raises an issue when at least one operand of an identity operator is a new object which has been created just for this check.</p>
<h2>Why is this an issue?</h2>
<p>Identity operators <code>is</code> and <code>is not</code> check if the same object is on both sides, i.e. <code>a is b</code> returns
<code>True</code> if <code>id(a) == id(b)</code>.</p>
<p>When a new object is created, it will have its own identity. Thus, if an object is created and used only in an identity check, it is not possible
for the other operand to be the same object. The comparison is always <code>False</code> or always <code>True</code> depending on the operator used,
<code>is</code> or <code>is not</code>.</p>
<p>This rule raises an issue when at least one operand of an identity operator is a new object which has been created just for this check, i.e.:</p>
<ul>
  <li> When it is a dict, list or set literal. </li>
  <li> When it is a call to <code>dict</code>, <code>set</code>, <code>list</code> or <code>complex</code> built-in functions. </li>
  <li> When such a new object is assigned to only one variable and this variable is used in an identity check. </li>
</ul>
<h2>How to fix it</h2>
<p>Whenever using a newly created object in a comparison, the identity operator should be replaced with the equality operator (<code>==</code> or
<code>!=</code>), which will use <code>__eq__</code> or <code>__ne__</code> methods under the hood.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre>
def func(param):
    param is {1: 2}  # Noncompliant: always False
    param is not {1, 2, 3}  # Noncompliant: always True
    param is [1, 2, 3]  # Noncompliant: always False

    param is dict(a=1)  # Noncompliant: always False

    mylist = []  # mylist is assigned a new object
    param is mylist  # Noncompliant: always False
</pre>
<h4>Compliant solution</h4>
<pre>
def func(param):
    param == {1: 2}
    param != {1, 2, 3}
    param == [1, 2, 3]

    param == dict(a=1)

    mylist = []
    param == mylist
</pre>
<h2>Resources</h2>
<h3>Articles &amp; blog posts</h3>
<ul>
  <li> <a href="https://adamj.eu/tech/2020/01/21/why-does-python-3-8-syntaxwarning-for-is-literal/">Why does Python 3.8 log a SyntaxWarning for 'is'
  with literals?</a> - Adam Johnson </li>
  <li> <a href="https://treyhunner.com/2019/03/unique-and-sentinel-values-in-python/#Equality_vs_identity">Equality vs identity</a> - Trey Hunner
  </li>
</ul>

