<p>This rule raises an issue when strings or bytes are concatenated implicitly.</p>
<h2>Why is this an issue?</h2>
<p>Python concatenates adjacent string or byte literals at compile time. It means that <code>"a" "b"</code> is equivalent to <code>"ab"</code>. This
is sometimes used to split a long string on multiple lines. However an implicit string concatenation can also be very confusing. In the following
contexts it might indicate that a comma was forgotten:</p>
<ul>
  <li> when the two strings are on the same line. This could be interpreted as an incorrectly formatted tuple (parentheses are not mandatory to create
  a tuple, only the comma is). </li>
  <li> when the strings are in a list, a set or a tuple. </li>
</ul>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
def func():
    return "item1" "item2"  # Noncompliant: a comma is missing to return a tuple.

["1"  # Noncompliant: a comma is missing.
 "2",
 "a very"  # Noncompliant: a "+" is missing.
 "long string"]
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
def func():
    return "item1", "item2"

["1",
 "2",
 "a very" +
 "long string"]
</pre>
<h3>Exceptions</h3>
<p>This rule will not raise any issues when there is a visible reason for the string concatenation:</p>
<ul>
  <li> when the quotes used for both strings are different. This can be used to avoid escaping quotes. </li>
  <li> when the strings or bytes have different prefixes, i.e. "f" for f-strings, "r" for raw, "u" for unicode and no prefix for normal strings. </li>
  <li> when strings are visibly split to avoid long lines of code, i.e. when the first string ends with a space, punctuation or <code>\n</code>. </li>
</ul>

