<p>A regex should never include a repetitive pattern whose body would match the empty string. This is usually a sign that a part of the regex is
redundant or does not do what the author intended.</p>
<h2>Noncompliant Code Example</h2>
<pre>
r"(?:)*"      # same as the empty regex, the '*' accomplishes nothing
r"(?:|x)*"    # same as the empty regex, the alternative has no effect
r"(?:x|)*"    # same as 'x*', the empty alternative has no effect
r"(?:x*|y*)*" # same as 'x*', the first alternative would always match, y* is never tried
r"(?:x?)*"    # same as 'x*'
r"(?:x?)+"    # same as 'x*'
</pre>
<h2>Compliant Solution</h2>
<pre>
r"x*"
</pre>

