<p>Why use named groups only to never use any of them later on in the code?</p>
<p>This rule raises issues every time named groups are:</p>
<ul>
  <li> referenced while not defined. </li>
</ul>
<h2>Noncompliant Code Example</h2>
<pre>
import re

def foo():
    pattern = re.compile(r"(?P&lt;a&gt;.)")
    matches = pattern.match("abc")
    g = matches.group("b") # Noncompliant - group "b" is not defined
    g = matches.group(2) # Noncompliant - there is no group with this number
</pre>
<h2>Compliant Solution</h2>
<pre>
import re

def foo():
    pattern = re.compile(r"(?P&lt;a&gt;.)")
    matches = pattern.match("abc")
    g = matches.group("a")
    g = matches.group(1)
</pre>

