<h2>Why is this an issue?</h2>
<p>Type hints can be used to communicate the intended type of a given variable. These are not enforced at runtime and not respecting them might not
necessarily lead to runtime errors.</p>
<p>It is however confusing and could lead to maintainability issues.</p>
<h3>Noncompliant code example</h3>
<pre>
def my_function():
    my_int: int = "string"  # Noncompliant
</pre>
<h3>Compliant solution</h3>
<pre>
def my_function():
    my_str: str = "string"
</pre>

