<h2>Why is this an issue?</h2>
<p>The <code>unittest</code> module provides assertion methods specific to common types and operations. Both versions will test the same things, but
the dedicated one will provide a better error message, simplifying the debugging process.</p>
<p>This rule reports an issue when an assertion can be simplified by using a more specific function. The array below gives a list of assertions on
which an issue will be raised, and which function should be used instead:</p>
<p>||Original||Dedicated||</p>
<p>|<code>assertTrue(x == y)</code>|<code>assertEqual(x, y)</code>|</p>
<p>|<code>assertTrue(x != y)</code>|<code>assertNotEqual(x, y)</code>|</p>
<p>|<code>assertFalse(x == y)</code>|<code>assertNotEqual(x, y)</code>|</p>
<p>|<code>assertFalse(x != y)</code>|<code>assertEqual(x, y)</code>|</p>
<p>|<code>assertTrue(x &lt; y)</code>|<code>assertLess(x, y)</code>|</p>
<p>|<code>assertTrue(x &lt;= y)</code>|<code>assertLessEqual(x, y)</code>|</p>
<p>|<code>assertTrue(x &gt; y)</code>|<code>assertGreater(x, y)</code>|</p>
<p>|<code>assertTrue(x &gt;= y)</code>|<code>assertGreaterEqual(x, y)</code>|</p>
<p>|<code>assertTrue(x is y)</code>|<code>assertIs(x, y)</code>|</p>
<p>|<code>assertTrue(x is not y)</code>|<code>assertIsNot(x, y)</code>|</p>
<p>|<code>assertFalse(x is y)</code>|<code>assertIsNot(x, y)</code>|</p>
<p>|<code>assertFalse(x is not y)</code>|<code>assertIs(x, y)</code>|</p>
<p>|<code>assertTrue(x in y)</code>|<code>assertIn(x, y)</code>|</p>
<p>|<code>assertFalse(x in y)</code>|<code>assertNotIn(x, y)</code>|</p>
<p>|<code>assertTrue(isinstance(x, y))</code>|<code>assertIsInstance(x, y)</code>|</p>
<p>|<code>assertFalse(isinstance(x, y))</code>|<code>assertNotIsInstance(x, y)</code>|</p>
<p>|<code>assertEqual(x, round(y, z))</code>|<code>assertAlmostEqual(x, y, z)</code>|</p>
<p>|<code>assertAlmostEqual(x, round(y, z))</code>|<code>assertAlmostEqual(x, y, z)</code>|</p>
<p>|<code>assertNotEqual(x, round(y, z))</code>|<code>assertNotAlmostEqual(x, y, z)</code>|</p>
<p>|<code>assertNotAlmostEqual(x, round(y, z))</code>|<code>assertNotAlmostEqual(x, y, z)</code>|</p>
<p>|<code>assertEqual(x, None)</code>|<code>assertIsNone(x)</code>|</p>
<p>|<code>assertNotEqual(x, None)</code>|<code>assertIsNotNone(x)</code>|</p>
<p>|<code>assertTrue(x is None)</code>|<code>assertIsNone(x)</code>|</p>
<p>|<code>assertTrue(x is not None)</code>|<code>assertIsNotNone(x)</code>|</p>
<p>|<code>assertFalse(x is None)</code>|<code>assertIsNotNone(x)</code>|</p>
<p>|<code>assertFalse(x is not None)</code>|<code>assertIsNone(x)</code>|</p>
<p>|<code>assertEqual(x, True)</code>|<code>assertTrue(x)</code>|</p>
<p>|<code>assertEqual(x, False)</code>|<code>assertFalse(x)</code>|</p>
<h3>Noncompliant code example</h3>
<pre>
import unittest
class SomeTest(unittest.TestCase):
  def test_something(self):
    x = foo()
    y = bar()
    self.assertFalse(x == y)  # Noncompliant
    self.assertTrue(x &lt; y)  # Noncompliant
</pre>
<h3>Compliant solution</h3>
<pre>
class SomeTest(unittest.TestCase):
  def test_something(self):
    x = foo()
    y = bar()
    self.assertNotEqual(x, y)
    self.assertLess(x, y)
</pre>
<h2>Resources</h2>
<p><a href="https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual">Python documentation - the <code>unittest</code>
module</a></p>

