<p>This rule raises an issue when an assertion result is constant.</p>
<h2>Why is this an issue?</h2>
<p>Assertions are meant to detect when code behaves as expected. An assertion which fails or succeeds all the time does not achieve this. Either it is
redundant and should be removed to improve readabity or it is a mistake and the assertion should be corrected.</p>
<p>This rule raises an issue when an assertion method is given parameters which will make it succeed or fail all the time. It covers three cases:</p>
<ul>
  <li> an <code>assert</code> statement or a unittest’s <code>assertTrue</code> or <code>assertFalse</code> method is called with a value which will
  be always True or always False. </li>
  <li> a unittest’s <code>assertIsNotNone</code> or <code>assertIsNone</code> method is called with a value which will always be None or never be
  None. </li>
  <li> a unittest’s <code>assertIsNot</code> or <code>assertIs</code> method is called with a literal expression creating a new object every time (ex:
  <code>[1, 2, 3]</code>). </li>
</ul>
<h3>Noncompliant code example</h3>
<pre data-diff-id="1" data-diff-type="noncompliant">
import unittest

class MyTestCase(unittest.TestCase):
    def expect_not_none(self):
        self.assertIsNotNone(round(1.5))  # Noncompliant: This assertion always succeeds because "round" returns a number, not None.

    def helper_compare(param):
        self.assertIs(param, [1, 2, 3])  # Noncompliant: This assertion always fails because [1, 2, 3] creates a new object.
</pre>
<h3>Compliant solution</h3>
<pre data-diff-id="1" data-diff-type="compliant">
import unittest

class MyTestCase(unittest.TestCase):
    def expect_not_none(self):
        self.assertNotEqual(round(1.5), 0)

    def helper_compare(param):
        self.assertEqual(param, [1, 2, 3])
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> <a href="https://docs.python.org/3/library/unittest.html">Python documentation - the <code>unittest</code> module</a> </li>
  <li> <a href="https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement">Python documentation - the <code>assert</code>
  statement</a> </li>
</ul>

