<h2>Why is this an issue?</h2>
<p>Using <code>pytest.raises</code> or <code>unittest.TestCase.assertRaises</code> will assert that an exception is raised in the following block.
Ending such block in an assertion means that the test can succeed with that last assertion never being executed.</p>
<h3>Noncompliant code example</h3>
<pre>
import pytest
def foo(): return 1 / 0
def bar(): return 42
def test_something():
  with pytest.raises(ZeroDivisionError):
    foo()
    assert bar() == 42  # Noncompliant
</pre>
<h3>Compliant solution</h3>
<pre>
import pytest
def foo(): return 1 / 0
def bar(): return 42
def test_something():
  with pytest.raises(ZeroDivisionError):
    foo()
  assert bar() == 42
</pre>
<h2>Resources</h2>
<ul>
  <li> <a href="https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises">Unittest: assertRaises</a> </li>
  <li> <a href="https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions">Pytest: assertions about expected exceptions</a>
  </li>
</ul>

