<h2>Why is this an issue?</h2>
<p>Variables, Classes and functions should not be undefined, otherwise the code will fail with a NameError.</p>
<h3>Noncompliant code example</h3>
<pre>
my_var # Noncompliant (variable is never defined)

def noncompliant():
    foo()  # Noncompliant
    MyClass()  # Noncompliant
</pre>
<h3>Compliant solution</h3>
<pre>
from mod import my_var

my_var

def compliant():
    foo = sum
    foo()

    class MyClass:
        pass
    MyClass()
</pre>

