<h2>Why is this an issue?</h2>
<p>Possessive quantifiers in Regex patterns like below improve performance by eliminating needless backtracking:</p>
<pre>
?+ , *+ , ++ , {n}+ , {n,}+ , {n,m}+
</pre>
<p>But because possessive quantifiers do not keep backtracking positions and never give back, the following sub-patterns should not match only similar
characters. Otherwise, possessive quantifiers consume all characters that could have matched the following sub-patterns and nothing remains for the
following sub-patterns.</p>
<h3>Noncompliant code example</h3>
<pre>
import re
pattern1 = re.compile(r"a++abc", re.DOTALL) # Noncompliant, the second 'a' never matches
pattern2 = re.compile(r"\d*+[02468]", re.DOTALL) # Noncompliant, the sub-pattern "[02468]" never matches
</pre>
<h3>Compliant solution</h3>
<pre>
import re
pattern1 = re.compile(r"aa++bc", re.DOTALL) # Compliant, for example it can match "aaaabc"
pattern2 = re.compile(r"\d*+(?&lt;=[02468])", re.DOTALL) # Compliant, for example, it can match an even number like "1234"
</pre>

