<p>When a back reference in a regex refers to a capturing group that hasn’t been defined yet (or at all), it can never be matched. Named back
references throw a <code>PatternSyntaxException</code> in that case; numeric back references fail silently when they can’t match, simply making the
match fail.</p>
<p>When the group is defined before the back reference but on a different control path (like in <code>(.)|\1</code> for example), this also leads to a
situation where the back reference can never match.</p>
<h2>Noncompliant Code Example</h2>
<pre>
import re
pattern = re.compile(r"\1(.)", re.DOTALL) # Noncompliant, group 1 is defined after the back reference
pattern = re.compile(r"(.)\2", re.DOTALL) # Noncompliant, group 2 isn't defined at all
pattern = re.compile(r"(.)|\1", re.DOTALL) # Noncompliant, group 1 and the back reference are in different branches
pattern = re.compile(r"(?P&lt;x&gt;.)|(?P=x)", re.DOTALL) # Noncompliant, group x and the back reference are in different branches
</pre>
<h2>Compliant Solution</h2>
<pre>
import re
pattern = re.compile(r"(.)\1", re.DOTALL)
pattern = re.compile(r"(?P&lt;x&gt;.)(?P=x)", re.DOTALL)
</pre>

