<h2>Why is this an issue?</h2>
<p>Python 3.11 introduced <code>except*</code> and <code>ExceptionGroup</code>. While itâ€™s possible to catch the ExceptionGroup and BaseExceptionGroup
types with <code>except`</code>, a Runtime error will be raised when this is done with <code>except*</code>. See PEP 654 : <a
href="https://peps.python.org/pep-0654/#forbidden-combinations">PEP-654</a></p>
<h3>Noncompliant code example</h3>
<pre>
try:
    ...
except* ExceptionGroup:  # Noncompliant
    pass

try:
    ...
except* (TypeError, ExceptionGroup):  # Noncompliant
    pass
</pre>
<h3>Compliant solution</h3>
<pre>
try:
    ...
except ExceptionGroup:
    pass
</pre>
<h2>Resources</h2>
<p><a href="https://peps.python.org/pep-0654/#forbidden-combinations">PEP-654</a></p>

