<h2>Why is this an issue?</h2>
<p>Python 3.10 introduced a specific syntax using the "or" operator (<code>X | Y</code>) to represent a union of types. This syntax has the same
functionality as <code>typing.Union</code>, but it is more concise and easier to read.</p>
<p>Using <code>typing.Union</code> is more verbose and less convenient. It can also create inconsistencies when different parts of the codebase use
different syntaxes for the same type.</p>
<h2>How to fix it</h2>
<p>Replace usages of <code>typing.Union</code> with the union type syntax.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre>
from typing import Union

def foo(arg: Union[int, str]) -&gt; Union[int, str]:
    if isinstance(arg, int):
        return arg + 1
    else:
        return arg.upper()
</pre>
<h4>Compliant solution</h4>
<pre>
def foo(arg: int | str) -&gt; int | str:
    if isinstance(arg, int):
        return arg + 1
    else:
        return arg.upper()
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> <a href="https://peps.python.org/pep-0604/">PEP 604 - Allow writing union types as X | Y</a> </li>
  <li> <a href="https://docs.python.org/3/library/stdtypes.html#types-union">Python documentation on union type expressions</a> </li>
</ul>

