<p>This rule raises an issue when a generator is passed to <code>np.array</code>.</p>
<h2>Why is this an issue?</h2>
<p>The creation of a NumPy array can be done in several ways, for example by passing a Python list to the <code>np.array</code> function. Another way
would be to pass a generator to the <code>np.array</code> function, but doing so creates a 0-dimensional array of objects and may not be the intended
goal. This NumPy array will have a have a data type (dtype) of <code>object</code> and could hold any Python objects.</p>
<p>One of the characteristics of NumPy arrays is homogeneity, meaning all its elements are of the same type. Creating an array of objects allows the
user to create heterogeneous array without raising any errors and creating such an array can lead to bugs further in the program.</p>
<pre data-diff-id="1" data-diff-type="noncompliant">
arr = np.array(x**2 for x in range(10))

arr.reshape(1)
arr.resize(2)
arr.put(indices=1, values=3) # No issues raised.
</pre>
<p>The NumPy array <code>arr</code> shown above now holds 2 values: a generator and the number 3.</p>
<h2>How to fix it</h2>
<p>To fix this issue, either:</p>
<ul>
  <li> pass a Python list instead of a generator to the <code>np.array</code> function or, </li>
  <li> explicitly show the intention to create a 0-dimensional array of objects by either adding <code>Any</code> as the type hint of the generator or
  by specifying the <code>dtype</code> parameter of the NumPy array as <code>object</code>. </li>
</ul>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="2" data-diff-type="noncompliant">
arr = np.array(x**2 for x in range(10)) # Noncompliant: the resulting array will be of the data type: object.

gen = (x*2 for x in range(5))
arr = np.array(gen) # Noncompliant: the resulting array will be of the data type: object.
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="2" data-diff-type="compliant">
from typing import Any

arr = np.array([x**2 for x in range(10)]) # Compliant: a list of 10 elements is passed to the np.array function.

arr = np.array(x**2 for x in range(10), dtype=object) # Compliant: the dtype parameter of np.array is set to object.

gen: Any = (x*2 for x in range(5))
arr = np.array(gen) # Compliant: the generator is explicitly type hinted with Any.
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> NumPy Documentation - <a href="https://numpy.org/doc/stable/reference/typing.html#arraylike">ArrayLike</a> </li>
</ul>

