<p>This rule raises an issue when a <code>TypeVar</code> is used in a function instead of the generic function type syntax.</p>
<h2>Why is this an issue?</h2>
<p>Prior to Python 3.12 functions using generic types were created as follows:</p>
<pre>
from typing import TypeVar

_T = TypeVar("_T")

def func(a: _T, b: _T) -&gt; _T:
    ...
</pre>
<p>This snippet of code can be confusing and difficult to read. This is why is it a good idea to use the new type parameter syntax of Python 3.12 that
allows for a more concise and more readable code by removing the <code>TypeVar</code> and its import statement:</p>
<pre>
def func[T](a: T, b: T) -&gt; T:
    ...
</pre>
<h3>Exceptions</h3>
<p>This rule will only raise an issue when the Python version of the analyzed project is set to 3.12 or higher.</p>
<h2>How to fix it</h2>
<p>To fix this issue, replace the usage of a <code>TypeVar</code> as a parameter type with the generic type parameter syntax.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
from typing import TypeVar

_T = TypeVar("_T", bound=str)

def func(a: _T, b: int) -&gt; _T: # Noncompliant: the usage of a TypeVar could be simplified with a generic type parameter.
    ...
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
def func[T: str](a: T, b: int) -&gt; T: # Compliant: the code is clear and concise.
    ...
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Python Documentation - <a href="https://docs.python.org/3.12/reference/simple_stmts.html#type">The type statement</a> </li>
  <li> Python 3.12 Release Notes - <a href="https://docs.python.org/3.12/whatsnew/3.12.html#pep-695-type-parameter-syntax">PEP 695: Type Parameter
  Syntax</a> </li>
  <li> PEP 695 - <a href="https://peps.python.org/pep-0695/">Type Parameter Syntax</a> </li>
</ul>

