<p>This rule raises an issue when a <code>numpy.busday_offset</code> object is created with an incorrect weekmask.</p>
<h2>Why is this an issue?</h2>
<p>To allow a datetime to be used in contexts where only certain days of the week are valid, NumPy includes a set of business day functions.
<code>Weekmask</code> is used to customize valid business days.</p>
<p><code>Weekmask</code> can be specified in several formats:</p>
<ol>
  <li> As an array of 7 <code>1</code> or <code>0</code> values, e.g. <code>[1, 1, 1, 1, 1, 0, 0]</code> </li>
  <li> As a string of 7 <code>1</code> or <code>0</code> characters, e.g. <code>"1111100"</code> </li>
  <li> As a string with abbreviations of valid days from this list: <code>Mon Tue Wed Thu Fri Sat Sun</code>, e.g. <code>"Mon Tue Wed Thu Fri"</code>
  </li>
</ol>
<p>Setting an incorrect <code>weekmask</code> leads to <code>ValueError</code>.</p>
<h2>How to fix it</h2>
<p>Provide a <code>weekmask</code> with correct values.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
import numpy as np

offset = np.busday_offset('2012-05', 1, roll='forward', weekmask='01') # Noncompliant: ValueError
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
import numpy as np

offset = np.busday_offset('2012-05', 1, roll='forward', weekmask='0111100') # OK
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Numpy documentation - <a href="https://numpy.org/doc/stable/reference/arrays.datetime.html#business-day-functionality">Business Day
  Functionality</a> </li>
  <li> Numpy documentation - <a href="https://numpy.org/doc/stable/reference/arrays.datetime.html#custom-weekmasks">Custom Weekmasks</a> </li>
</ul>

