<p>This rule raises an issue when a <code>tensorflow.Variable</code> created inside of a <code>tensorflow.function</code> is not a singleton.</p>
<h2>Why is this an issue?</h2>
<p><code>tensorflow.function</code>s only supports singleton <code>tensorflow.Variable</code>s. This means the variable will be created on the first
call of the <code>tensorflow.function</code> and will be reused across the subsequent calls. Creating a <code>tensorflow.Variable</code> that is not a
singleton will raise a <code>ValueError</code>.</p>
<pre>
import tensorflow as tf

@tf.function
def f(x):
  v = tf.Variable(1.0)
  return v
</pre>
<p>In the example above each time the function <code>f</code> will be called a new <code>tensorflow.Variable</code> will be created.</p>
<h2>How to fix it</h2>
<p>To fix this issue use the singleton pattern to avoid the creation of a <code>tensorflow.Variable</code> multiple times.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
import tensorflow as tf

class Count(tf.Module):
  def __init__(self):
    self.count = None

  @tf.function
  def __call__(self):
    self.count = tf.Variable(0) # Noncompliant: the variable will be re-created for each call
    return self.count.assign_add(1)

c = Count()
c() # ValueError is raised
c()
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
import tensorflow as tf

class Count(tf.Module):
  def __init__(self):
    self.count = None

  @tf.function
  def __call__(self):
    if self.count is None:
      self.count = tf.Variable(0) # Compliant: the variable is created only on during the first call
    return self.count.assign_add(1)

c = Count()
c()
c()
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> TensorFlow Documentation - <a href="https://www.tensorflow.org/guide/function#creating_tfvariables">Creating tf.Variables</a> </li>
</ul>

