<p>This rule raises an issue when <code>pytorch.load</code> is used to load a model.</p>
<h2>Why is this an issue?</h2>
<p>In PyTorch, it is common to load serialized models using the <code>torch.load</code> function. Under the hood, <code>torch.load</code> uses the
<code>pickle</code> library to load the model and the weights. If the model comes from an untrusted source, an attacker could inject a malicious
payload which would be executed during the deserialization.</p>
<h2>How to fix it</h2>
<p>Use a safer alternative to load the model, such as <code>safetensors.torch.load_model</code>. Alternatively, PyTorch can be instructed to only load
the weights by setting the parameter <code>weights_only=True</code>. This avoids the use of the <code>pickle</code> library and is therefore safe.
Note that the use of <code>weights_only</code> requires saving only the <code>state_dict</code> of a model instead of the whole model.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
import torch

model = torch.load('model.pth') # Noncompliant: torch.load is used to load the model
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
import torch
import safetensors

model = MyModel()
safetensors.torch.load_model(model, 'model.pth')
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Pytorch documentation: <a href="https://pytorch.org/tutorials/beginner/saving_loading_models.html#save-load-entire-model">Save/Load Entire
  Model</a> </li>
</ul>

