<p>Statements with no side effects and no change of control flow do not contribute to the functionality of the code and can indicate a programming
error.</p>
<h2>Why is this an issue?</h2>
<p>When writing code, it is important to ensure that each statement serves a purpose and contributes to the overall functionality of the program. When
they have no side effects or do not change the control flow, they can either indicate a programming error or be redundant:</p>
<ol>
  <li> The code does not behave as intended: The statements are expected to have an effect but they do not. This can be caused by mistyping,
  copy-and-paste errors, etc. </li>
  <li> The statements are residual after a refactoring. </li>
</ol>
<h3>Exceptions</h3>
<p><strong>Intentionally empty statement</strong></p>
<p>Statements such as <code>pass</code> or <code>...</code> (ellipsis) are clearly meant to have no effect and may be used to indicate an
implementation is missing. No issue will be raised in this case.</p>
<p><strong>Strings</strong></p>
<p>Some projects use string literals as comments. By default, this rule will not raise an issue on these strings. Reporting on string literals can be
enabled by setting the rule parameter <code>reportOnStrings</code> to <code>true</code>.</p>
<pre>
def foo():
    bar()
    """Some comment"""  # Compliant by default. Noncompliant with "reportOnStrings" set to "true"
    qix()
</pre>
<p><strong>Operators</strong></p>
<p>By default, this rule considers that no arithmetic operator has a side effect. Some projects may redefine operators and add a side effect. You can
list such operators in the rule parameter <code>ignoredOperators</code>.</p>
<pre>
def process(p, beam):
    """
    Apache Beam redefines "|" and "&gt;&gt;" operators and they have a side effect.
    Thus for Apache Beam projects "ignoredOperators"should be set to "|,&gt;&gt;"
    """
    p | "create" &gt;&gt; beam.Create()  # Noncompliant by default
</pre>
<h2>How to fix it</h2>
<p>Identify statements that do not contribute to the functionality of the code and verify if they are intended to be part of the logic. If they are,
there is a bug to be fixed. If they are not, then they are redundant and should be removed.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre>
def get_result():
    result = 42
    if should_be_zero():
        result == 0 # Noncompliant: no side effect, was an assignment intended?
    return result
</pre>
<h4>Compliant solution</h4>
<pre>
def get_result():
    result = 42
    if should_be_zero():
        result = 0 # Compliant
    return result
</pre>
<h2>Resources</h2>
<h3>Standards</h3>
<ul>
  <li> CWE - <a href="https://cwe.mitre.org/data/definitions/482">482 Comparing instead of Assigning</a> </li>
</ul>

